### PR TITLE
Add MP0004 Fault Tolerance Proposal

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang6@googlemail.com>

# Proposal for a MicroProfile specification

**! DO NOT MERGE this Pull Request BEFORE REVIEW is finished!**

**Merging this Pull Request means that the proposal is officially accepted by the MicroProfile Core Team as a MicroProfile incubation project. A public discussion and an agreement within the MicroProfile Core Team is expected before a proposal is accepted/rejected.** 

_If this PR is not a proposal for a new specification, delete this template description._

## Suggested steps to follow before this proposal is accepted ##

 - add label `proposal` to this PR on github
 - initiate a public discussion on the mP mailing list
   - Email Subject: "Review announcement: <<PROPOSAL_NAME>>"
   - Email Body: Based on [review announcement template](https://github.com/microprofile/evolution/blob/master/process.md#review-announcement) 
       - fill in `<<PROPOSAL NAME>>`, `<<REVIEW END DATE>>` (not mandatory, can be announced later, but should be at least 7 days after the announcement email is sent) 
       and `<<REVIEW MANAGER NAME>>` (the name of the sender)
       - update the link to a proposal (NNNN-proposal.md) to point to this Pull Request

After the review, the Core Team should form an agreement whether or not to accept the proposal, or whether ask its author for amendments. 

Once this Pull Reuest is merged, the proposal is considered accepted and an incubation repository should be created to facilitate collaboration.

### CODE REVIEW

File: proposals/0004-FaultTolerance.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
+# Fault Tolerance
+
+* Proposal: [MP-0004](0004-FaultTolerance.md)
+* Authors: [Emily Jiang](https://github.com/Emily-Jiang), [Jonathan Halterman](https://github.com/jhalterman/)
```
Comment (johnament) [inline proposals/0004-FaultTolerance.md]: Please add Antoine and I.

Comment (Emily-Jiang) [reply proposals/0004-FaultTolerance.md]: added

**Comment by OndroMih** [comment]
This is a follow-up to closed #15. Look there for previous discussions.

**Comment by Pscheidl** [comment]
https://github.com/Pscheidl/FortEE

I was creating a fault-tolerance solution based on Java EE on my own. Currently, basic failsafe and absolute timeout are done. We plan to add at least circuit breaker (very easy and very soon) and a machine learning approach to timeouts.

Maybe you can have a look and if nothing else, you can see which way is the bad way to go :) It is still in Beta and we are trying to test it.

More information in README and also in Microprofile Google groups here: https://groups.google.com/forum/#!topic/microprofile/2XH_1ddVXcg

File: proposals/0004-FaultTolerance.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!topic/microprofile/ezFC1TLGozU), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+It is increasingly important to build fault tolerant micro services. Fault tolerance is about leveraging different strategies to guide the execution and result of some logic. Retry policies, bulkheads, and circuit breakers are popular concepts in this area. They dictate whether and when executions should take place, and fallbacks offer an alternative result when an execution does not complete successfully. 
+
+As mentioned above, the Fault Tolerance proposal is to focus the aspects: TimeOut, RetryPolicy, Fallback, bulkhead and circuit breaker.
+
+* TimeOut: Define a duration for timeout
```
Comment (Pscheidl) [inline proposals/0004-FaultTolerance.md:17]: Absolute duration for timeout might be unsuable. A confidence interval can possibly reduce waste of system resources in ever-changing world where latencies are highly unstable.

File: proposals/0004-FaultTolerance.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+* CDI-first approach to apply RetryPolicy, Fallback, BulkHead, CircuitBreaker using annotations
+
+## Detailed design (One example of implemenations)
+
+
+### CDI-based approach 
+Use interceptor and annotation to specify the execution and policy configration.
+An annotation of Asynchronous has to be specified for any asynchronous calls. Otherwise, synchronous execution is assumed. 
+The implementation should provide two interceptors, one for synchronous invocation and the other for asynchronous invocation.
```
Comment (Pscheidl) [inline proposals/0004-FaultTolerance.md:62]: If synchronous and asynchronous execution is desired in one bean, does this imply a need for double injection ? Once with the async interceptor and once with the sync interceptor ?

File: proposals/0004-FaultTolerance.md (hunk 2 above)
Comment (Emily-Jiang) [reply proposals/0004-FaultTolerance.md:17]: I will merge this proposal and we can discuss the APIs/SPIs once the Fault Tolerance apis are created.

File: proposals/0004-FaultTolerance.md (hunk 3 above)
Comment (Emily-Jiang) [reply proposals/0004-FaultTolerance.md:62]: not apply both. The implementation provides both. The clients only choose the one they want.

File: proposals/0004-FaultTolerance.md (hunk 2 above)
Comment (Emily-Jiang) [reply proposals/0004-FaultTolerance.md:17]: Thanks for sharing @Pscheidl ! I will look at your impl. Please join the Fault Tolerance hangout and discuss the details.

Comment (Pscheidl) [reply proposals/0004-FaultTolerance.md:17]: Roger that, will do. I just finished investigating the timeout possibilities and we can see many options. Will describe those options in the discussion and you can see if there is something useful for spec, or it is impl detail.

**Pull request merged**